### ISSUE TITLE
Potential deadlock in rcache?

### ISSUE THREAD

**Issue by devreal** [open-mpi/ompi#7303]
I'm trying to run a PaRSEC benchmark (`testing_dgeqrf`) on a Cray XC40 with the latest v4.0.x compiled with the Intel compiler. For larger problem sizes (16 nodes, 24 threads per node, 20k^2 elements per process) the run locks up after all tasks have been processed and I see the following aggregated backtrace in DDT:

```
Threads,Function
1,progress_engine (opal_progress_threads.c:105)
1,  opal_libevent2022_event_base_loop (event.c:1630)
1,    poll_dispatch (poll.c:165)
1,      poll
1,progress_engine (pmix_progress_threads.c:232)
1,  opal_libevent2022_event_base_loop (event.c:1630)
1,    epoll_dispatch (epoll.c:407)
1,      epoll_wait
22,__parsec_thread_init
22,  __parsec_context_wait
18,    nanosleep
1,    sched_lfq_select
1,      parsec_hbbuffer_pop_best
3,    __parsec_task_progress
3,      complete_hook_of_dgeqrf_dtsmqr
3,        intercept_munmap (memory_patcher_component.c:167)
3,          _intercept_munmap (memory_patcher_component.c:153)
3,            opal_mem_hooks_release_hook (memory.c:129)
3,              mca_rcache_base_mem_cb (rcache_base_mem_cb.c:63)
3,                mca_rcache_grdma_invalidate_range (rcache_grdma_module.c:496)
3,                  mca_rcache_base_vma_iterate (rcache_base_vma.c:153)
3,                    mca_rcache_base_vma_tree_iterate (rcache_base_vma_tree.c:105)
3,                      opal_interval_tree_traverse (opal_interval_tree.c:706)
3,                        inorder_traversal (opal_interval_tree.c:670)
3,                          inorder_traversal (opal_interval_tree.c:654)
3,                            inorder_traversal (opal_interval_tree.c:670)
3,                              inorder_traversal (opal_interval_tree.c:670)
3,                                inorder_traversal (opal_interval_tree.c:670)
3,                                  inorder_traversal (opal_interval_tree.c:670)
3,                                    inorder_traversal (opal_interval_tree.c:670)
3,                                      inorder_traversal (opal_interval_tree.c:670)
3,                                        inorder_traversal (opal_interval_tree.c:670)
2,                                          inorder_traversal (opal_interval_tree.c:654)
1,                                            inorder_traversal (opal_interval_tree.c:654)
1,                                              inorder_traversal (opal_interval_tree.c:670)
1,                                                inorder_traversal (opal_interval_tree.c:654)
1,                                                  inorder_traversal (opal_interval_tree.c:664)
1,                                                    mca_rcache_base_vma_tree_iterate_helper (rcache_base_vma_tree.c:95)
1,                                                      gc_add (rcache_grdma_module.c:488)
1,                                                        mca_rcache_grdma_add_to_gc (rcache_grdma_module.c:461)
1,                                                          mca_rcache_grdma_remove_from_lru (rcache_grdma_module.c:232)
1,                                                            opal_mutex_lock (mutex_unix.h:137)
1,                                                              pthread_mutex_lock
1,                                                                __lll_lock_wait
1,                                            inorder_traversal (opal_interval_tree.c:670)
1,                                              inorder_traversal (opal_interval_tree.c:654)
1,                                                inorder_traversal (opal_interval_tree.c:654)
1,                                                  inorder_traversal (opal_interval_tree.c:664)
1,                                                    mca_rcache_base_vma_tree_iterate_helper (rcache_base_vma_tree.c:95)
1,                                                      gc_add (rcache_grdma_module.c:488)
1,                                                        mca_rcache_grdma_add_to_gc (rcache_grdma_module.c:461)
1,                                                          mca_rcache_grdma_remove_from_lru (rcache_grdma_module.c:232)
1,                                                            opal_mutex_lock (mutex_unix.h:137)
1,                                                              pthread_mutex_lock
1,                                                                __lll_lock_wait
1,                                          inorder_traversal (opal_interval_tree.c:670)
1,                                            inorder_traversal (opal_interval_tree.c:664)
1,                                              mca_rcache_base_vma_tree_iterate_helper (rcache_base_vma_tree.c:95)
1,                                                gc_add (rcache_grdma_module.c:488)
1,                                                  mca_rcache_grdma_add_to_gc (rcache_grdma_module.c:461)
1,                                                    mca_rcache_grdma_remove_from_lru (rcache_grdma_module.c:232)
1,                                                      opal_mutex_lock (mutex_unix.h:137)
1,                                                        pthread_mutex_lock
1,                                                          __lll_lock_wait
1,remote_dep_dequeue_main
1,  remote_dep_dequeue_nothread_progress
1,    remote_dep_mpi_save_put_cb
1,      PMPI_Isend (pisend.c:95)
1,        mca_pml_ob1_isend (pml_ob1_isend.c:198)
1,          mca_pml_ob1_send_request_start_seq (pml_ob1_sendreq.h:465)
1,            mca_pml_ob1_send_request_start_btl (pml_ob1_sendreq.h:419)
1,              mca_pml_ob1_rdma_btls (pml_ob1_rdma.c:89)
1,                mca_btl_ugni_register_mem (btl_ugni_module.c:309)
1,                  mca_rcache_grdma_register (rcache_grdma_module.c:349)
1,                    mca_rcache_grdma_evict (rcache_grdma_module.c:195)
1,                      mca_rcache_grdma_evict_lru_local (rcache_grdma_module.c:185) // <- the lock is taken here
1,                        dereg_mem (rcache_grdma_module.c:144)
1,                          mca_rcache_base_vma_delete (rcache_base_vma.c:145)
1,                            mca_rcache_base_vma_tree_delete (rcache_base_vma_tree.c:126)
1,                              opal_interval_tree_delete (opal_interval_tree.c:560)
1,                                opal_interval_tree_delete_interior (opal_interval_tree.c:509)
1,                                  rp_free_wait (opal_interval_tree.c:441)
1,                                    rp_wait_for_readers (opal_interval_tree.c:434)
1,main
1,  parsec_context_wait
1,    nanosleep
```

It looks to me like a deadlock situation where one thread holds a lock waiting for some condition that can only be fulfilled by other threads waiting for the same lock. The `cache->vma_module->vma_lock` is taken by the lower thread in `mca_rcache_grdma_evict_lru_local` before waiting for a condition in `rp_wait_for_readers`. The upper threads try to acquire that lock too down the road from `opal_mem_hooks_release_hook`, which I guess are the readers here? I am not familiar with the rcache code but it looks suspicious to me. Can someone please a look at this?

I'm fairly sure that this is not related to #7301.

**Comment by gpaulsen** [comment]
I wonder if this is also related to c11-atomics as mentioned in #7301?

**Comment by devreal** [comment]
@gpaulsen I'm fairly confident that the issue is unrelated to #7301 as it also occurs on the v4.0.x branch (#7301 does not) and is independent of the compiler (#7301 does not occur with the GNU compiler). I looked at it more closely with DDT and I strongly believe that it is a deadlock situation caused by the waiting thread holding the `vma_lock` while the threads it waits on try to acquire the lock before signalling the condition the first thread is waiting on in `rp_wait_for_readers`. #7307 reduces the scope in which the waiting thread holds the lock, which resolves the issue.

**Comment by hjelmn** [comment]
Plan to review that change early next week.

**Comment by hppritcha** [comment]
@gpaulsen if we can get #7307 PR'd over to 4.0.x prior to 4.0.3 release this would be nice to have for Cray's.

**Comment by gpaulsen** [comment]
Observations in #7319 make it appear that we should now hold off on accepting PR #7307 in v4.0.3 until we gather more data.

**Comment by hjelmn** [comment]
PR #7303 is not quite right. There is a reason the lock is not dropped there. Working on a fix.

**Comment by hjelmn** [comment]
I think I have a fix. @devreal If I give you a patch can you try it and see if it works for you?

**Comment by devreal** [comment]
Absolutely, I'll be happy to give it a try :+1:

**Comment by devreal** [comment]
@hjelm Can you please send me the patch to try? I'm still struggling with this issue...

**Comment by hjelmn** [comment]
http://github.com/hjelmn/ompi/tree/fix_bug_7303_the_rcache_deadlock

**Comment by devreal** [comment]
Thanks for the patch! I tested the branch, which gives me a bunch of warnings and a Segfault with the reproducer posted over at https://github.com/open-mpi/ompi/issues/7319 (running on 16 nodes, one rank per node, 24 threads each):

```
 Warning :: opal_list_remove_item - the item 0x8b6840 is not on the list 0x8b1a00 
 Warning :: opal_list_remove_item - the item 0x8b6640 is not on the list 0x8b1a00 
```

````
#9 main._omp_fn () at src/test/omp_mpi_send.c:37 (at 0x0000000000400e2a)
#8 PMPI_Isend (buf=0x8cf4f0, count=16384, type=0x6020c0, dest=12, tag=7, comm=0x6022c0, request=0x2aaad0000980) at build/ompi/mpi/c/profile/pisend.c:95 (at 0x00002aaaaad86442)
#7 mca_pml_ob1_isend (buf=0x8cf4f0, count=16384, datatype=0x6020c0, dst=12, tag=7, sendmode=MCA_PML_BASE_SEND_STANDARD, comm=0x6022c0, request=0x2aaad0000980) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_isend.c:198 (at 0x00002aaabd50a7bb)
#6 mca_pml_ob1_send_request_start_seq (sendreq=0x2aaadc05ae00, endpoint=0x8bc670, seqn=147) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_sendreq.h:464 (at 0x00002aaabd509bcf)
#5 mca_pml_ob1_send_request_start_btl (sendreq=0x2aaadc05ae00, bml_btl=0x8bc7d0) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_sendreq.h:418 (at 0x00002aaabd5099f3)
#4 mca_pml_ob1_rdma_btls (bml_endpoint=0x8bc670, base=0x8cf4f0 "", size=65536, rdma_btls=0x2aaadc05b158) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_rdma.c:89 (at 0x00002aaabd50cba2)
#3 mca_btl_ugni_register_mem (btl=0x7c6260, endpoint=0x897e50, base=0x8cf4f0, size=65536, flags=2) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_module.c:309 (at 0x00002aaabca2b4c9)
#2 mca_rcache_grdma_register (rcache=0x8b2650, addr=0x8cf4f0, size=65536, flags=0, access_flags=2, reg=0x2aaac0203ba8) at build/opal/mca/rcache/grdma/../../../../../opal/mca/rcache/grdma/rcache_grdma_module.c:317 (at 0x00002aaab6877c92)
#1 do_unregistration_gc (rcache=0x8b2650) at build/opal/mca/rcache/grdma/../../../../../opal/mca/rcache/grdma/rcache_grdma_module.c:204 (at 0x00002aaab6877914)
#0 dereg_mem (reg=0x8b1a48) at build/opal/mca/rcache/grdma/../../../../../opal/mca/rcache/grdma/rcache_grdma_module.c:178 (at 0x00002aaab68777f1)
````

The Segfault is caused by `rcache_grdma` having the value `0xdeafbeeddeafbeed`.

In a different place, I get the following Segfault:

```
#30 main._omp_fn () at src/test/omp_mpi_send.c:42 (at 0x0000000000400d7e)
#29 PMPI_Waitall (count=768, requests=0x2aab0c0008c0, statuses=0x0) at build/ompi/mpi/c/profile/pwaitall.c:80 (at 0x00002aaaaadadc66)
#28 ompi_request_default_wait_all (count=768, requests=0x2aab0c0008c0, statuses=0x0) at build/ompi/../../ompi/request/req_wait.c:243 (at 0x00002aaaaad3449a)
#27 ompi_sync_wait_mt (sync=0x2aaabfc00d20) at build/opal/../../opal/threads/wait_sync.c:88 (at 0x00002aaaabbfe049)
#26 opal_progress () at build/opal/../../opal/runtime/opal_progress.c:230 (at 0x00002aaaabbf4e82)
#25 mca_btl_ugni_component_progress () at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_component.c:721 (at 0x00002aaabca28657)
#24 mca_btl_ugni_progress_datagram (device=0x7c64a0) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_component.c:613 (at 0x00002aaabca27fcd)
#23 mca_btl_ugni_smsg_process (ep=0x896b20) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_smsg.c:114 (at 0x00002aaabca38af5)
#22 mca_pml_ob1_recv_frag_callback_rget (btl=0x7c6260, tag=67 'C', des=0x2aaabfc00960, cbdata=0x0) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_recvfrag.c:551 (at 0x00002aaabd510523)
#21 mca_pml_ob1_recv_frag_match (btl=0x7c6260, hdr=0x2aaabf6fdecc, segments=0x2aaabfc00950, num_segments=1, type=67) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_recvfrag.c:980 (at 0x00002aaabd51136d)
#20 mca_pml_ob1_recv_frag_match_proc (btl=0x7c6260, comm_ptr=0x6022c0, proc=0x2aab000020d0, hdr=0x2aaabf6fdecc, segments=0x2aaabfc00950, num_segments=1, type=67, frag=0x0) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_recvfrag.c:1046 (at 0x00002aaabd5114a6)
#19 mca_pml_ob1_recv_request_progress_rget (recvreq=0x2aab000127c0, btl=0x7c6260, segments=0x2aaabfc00950, num_segments=1) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_recvreq.c:792 (at 0x00002aaabd5158b7)
#18 mca_pml_ob1_recv_request_get_frag (frag=0x2aaad802f540) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_recvreq.c:498 (at 0x00002aaabd515217)
#17 mca_bml_base_get (bml_btl=0x8bab20, local_address=0x8cf3a0, remote_address=9237408, local_handle=0x8b5410, remote_handle=0x2aaad802f628, size=65536, flags=0, order=255, cbfunc=0x2aaabd514d20 <mca_pml_ob1_rget_completion>, cbdata=0x2aaad802f540) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/bml/bml.h:326 (at 0x00002aaabd51322c)
#16 mca_btl_ugni_get (btl=0x7c6260, endpoint=0x896b20, local_address=0x8cf3a0, remote_address=9237408, local_handle=0x8b5410, remote_handle=0x2aaad802f628, size=65536, flags=0, order=255, cbfunc=0x2aaabd514d20 <mca_pml_ob1_rget_completion>, cbcontext=0x8bab20, cbdata=0x2aaad802f540) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_get.c:43 (at 0x00002aaabca36f57)
#15 mca_btl_ugni_post (cbdata=0x2aaad802f540, cbcontext=0x8bab20, cbfunc=0x2aaabd514d20 <mca_pml_ob1_rget_completion>, order=255, remote_handle=0x2aaad802f628, local_handle=0x8b5410, remote_address=9237408, local_address=0x8cf3a0, size=65536, get=1, endpoint=0x896b20) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_rdma.h:145 (at 0x00002aaabca36f57)
#14 mca_btl_ugni_post_bte (endpoint=0x896b20, op_type=GNI_POST_RDMA_GET, size=65536, local_address=0x8cf3a0, remote_address=9237408, local_handle=0x8b5410, remote_handle=0x2aaad802f628, order=255, cbfunc=0x2aaabd514d20 <mca_pml_ob1_rget_completion>, cbcontext=0x8bab20, cbdata=0x2aaad802f540) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_rdma.h:101 (at 0x00002aaabca365ff)
#13 mca_btl_ugni_endpoint_post_rdma (desc=0x2aaabfbffec0, endpoint=0x896b20) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_device.h:533 (at 0x00002aaabca365ff)
#12 mca_btl_ugni_endpoint_post (post_fn=0x2aaabca35eb0 <mca_btl_ugni_post_rdma_device>, desc=0x2aaabfbffec0, endpoint=0x896b20) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_device.h:518 (at 0x00002aaabca365ff)
#11 mca_btl_ugni_handle_rdma_completions (ugni_module=0x7c6260, device=0x7c64a0, post_desc=0x2aaabfbffaf8, count=1) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_component.c:639 (at 0x00002aaabca28192)
#10 mca_btl_ugni_post_desc_complete (module=0x7c6260, desc=0x2aaabfbffaf8, rc=0) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_frag.h:187 (at 0x00002aaabca25a6a)
#9 mca_pml_ob1_rget_completion (btl=0x7c6260, ep=0x896b20, local_address=0x8cf3a0, local_handle=0x8b5410, context=0x8bab20, cbdata=0x2aaad802f540, status=0) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_recvreq.c:406 (at 0x00002aaabd514ee5)
#8 mca_pml_ob1_send_request_process_pending (bml_btl=0x8bab20) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_sendreq.c:77 (at 0x00002aaabd5196fe)
#7 mca_pml_ob1_send_request_start_btl (sendreq=0x2aaad8017b00, bml_btl=0x8bb7b0) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_sendreq.h:418 (at 0x00002aaabd519551)
#6 mca_pml_ob1_rdma_btls (bml_endpoint=0x8bb650, base=0x8cf3a0 "", size=65536, rdma_btls=0x2aaad8017e58) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_rdma.c:89 (at 0x00002aaabd50cba2)
#5 mca_btl_ugni_register_mem (btl=0x7c6260, endpoint=0x897360, base=0x8cf3a0, size=65536, flags=2) at build/opal/mca/btl/ugni/../../../../../opal/mca/btl/ugni/btl_ugni_module.c:309 (at 0x00002aaabca2b4c9)
#4 mca_rcache_grdma_register (rcache=0x8b2650, addr=0x8cf3a0, size=65536, flags=0, access_flags=2, reg=0x2aaabfbff868) at build/opal/mca/rcache/grdma/../../../../../opal/mca/rcache/grdma/rcache_grdma_module.c:317 (at 0x00002aaab6877c92)
#3 do_unregistration_gc (rcache=0x8b2650) at build/opal/mca/rcache/grdma/../../../../../opal/mca/rcache/grdma/rcache_grdma_module.c:203 (at 0x00002aaab6877908)
#2 mca_rcache_grdma_remove_from_lru (rcache_grdma=0x8b2650, grdma_reg=0x8b6540) at build/opal/mca/rcache/grdma/../../../../../opal/mca/rcache/grdma/rcache_grdma_module.c:162 (at 0x00002aaab6877756)
#1 opal_list_remove_item (list=0x8b1a20, item=0x8b6540) at build/opal/mca/rcache/grdma/../../../../../opal/class/opal_list.h:485 (at 0x00002aaab68768e2)
#0 opal_list_get_end (list=0x8b1a20) at build/opal/mca/rcache/grdma/../../../../../opal/class/opal_list.h:404 (at 0x00002aaab6876882)
```

**Comment by hjelmn** [comment]
Ok. Think I see the issue. Will fix and let you know when you can try again.

**Comment by hjelmn** [comment]
@devreal Try http://github.com/hjelmn/ompi/tree/fix_bug_7303_the_rcache_deadlock_v3  . Different approach but will likely work. Tested @ 64 KNL threads with a RMA-MT by adding per-round allocation/deallocation to the code. Seems to be working ok.

**Comment by hjelmn** [comment]
Spoke too soon. Found  a bug.

**Comment by hjelmn** [comment]
Now give it a try.

**Comment by devreal** [comment]
Thanks a lot! It seems that I can run the PaRSEC benchmark successfully with this patch :+1: 

However, the OpenMP benchmark posted in https://github.com/open-mpi/ompi/issues/7319 still throws an error:

```
omp_mpi_send: ../../../../../ompi/mca/pml/ob1/pml_ob1_sendreq.h:219: mca_pml_ob1_send_request_fini: Assertion `((void *)0) == sendreq->rdma_frag' failed.
```

```
#9 gomp_thread_start (xdata=<optimized out>) at /b/tmp/peint/build-cray-gcc-20190312.154136.620938870/cray-gcc/BUILD/snos_objdir/x86_64-suse-linux/libgomp/../../../cray-gcc-8.3.0-201903122028.16ea96cb84a9a/libgomp/team.c:120 (at 0x00002aaaab0be4be)
#8 main._omp_fn () (at 0x0000000000400d69)
#7 PMPI_Waitall (count=512, requests=0x2aab180008c0, statuses=0x0) at build/ompi/mpi/c/profile/pwaitall.c:80 (at 0x00002aaaaada9724)
#6 ompi_request_default_wait_all (count=512, requests=0x2aab180008c0, statuses=0x0) at build/ompi/../../ompi/request/req_wait.c:362 (at 0x00002aaaaad31a68)
#5 ompi_request_free (request=0x2aab18001750) at build/ompi/../../ompi/request/request.h:380 (at 0x00002aaaaad30e8f)
#4 mca_pml_ob1_send_request_free (request=0x2aab18001750) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_sendreq.c:120 (at 0x00002aaabd5191a0)
#3 mca_pml_ob1_send_request_fini (sendreq=0x2aaaec029e80) at build/ompi/mca/pml/ob1/../../../../../ompi/mca/pml/ob1/pml_ob1_sendreq.h:219 (at 0x00002aaabd518b65)
#2 __assert_fail () from /lib64/libc.so.6 (at 0x00002aaaab520e22)
#1 __assert_fail_base () from /lib64/libc.so.6 (at 0x00002aaaab520d76)
#0 abort () from /lib64/libc.so.6 (at 0x00002aaaab529210)
```

That might be an unrelated issue though.

**Comment by hjelmn** [comment]
@devreal I think that is unrelated. I can't reproduce the issue with my branch with PrgEnv-gnu running with ```mpirun --np 2 --bind-to socket --npernode 1```. Am I running it correctly?

**Comment by devreal** [comment]
Have you tried running it on more than two nodes? Your run command seems right. It appears for me on 16 nodes with a run command like this:

```
OMP_NUM_THREADS=24 mpirun -n 16 --bind-to none -N 1
```

**Comment by gpaulsen** [comment]
@hjelmn Do you know what's needed here?

**Comment by jsquyres** [comment]
@devreal Is this still an issue, or did it get fixed?

**Comment by devreal** [comment]
I cannot reproduce on that system because that system is gone. I will try again with `ob1/uct` once our new system is in production (1-2 weeks from now) and I will update/close this issue then.

**Comment by devreal** [comment]
I cannot test this anymore. Will close it and get back to it if I encounter it again.
